navigation/libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/instruction/DefaultInstructionEngine.java

```diff
@@ -0,0 +1,163 @@
+
+  private static final double MINIMUM_UPCOMING_STEP_DISTANCE = 15d;
+  private static final String DECIMAL_FORMAT = "###.#";
+  private static final String MILES_STRING_FORMAT = "%s Miles";
+  private static final String FEET_STRING_FORMAT = "%s Feet";
+  private static final String IN_STRING_FORMAT = "In %s %s";
+  private static final String THEN_STRING_FORMAT = "%s then %s";
+  private static final String THEN_IN_STRING_FORMAT = "%s then in %s %s";
+  private static final String CONTINUE_STRING_FORMAT = "Continue on %s for %s";
+
+  DefaultInstructionEngine() {
+    initDefaultBuilders();
+  }
+
+  private void initDefaultBuilders() {
+    this.put(NavigationConstants.DEPARTURE_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildDepartureInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.NEW_STEP_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildNewStepInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.IMMINENT_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildImminentInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.URGENT_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        buildUrgentInstruction(routeProgress);
+        return null;
+      }
+    });
+    this.put(NavigationConstants.ARRIVAL_MILESTONE, new InstructionBuilder() {
+      @Override
+      public String build(RouteProgress routeProgress) {
+        return buildArrivalInstruction(routeProgress);
+      }
+    });
+  }
+
+  private String buildDepartureInstruction(RouteProgress progress) {
+    if (progress.getCurrentLegProgress().getUpComingStep().getDistance() > MINIMUM_UPCOMING_STEP_DISTANCE) {
+      return buildContinueFormatInstruction(progress);
+    } else {
+      return buildThenInFormatInstruction(progress);
+    }
+  }
+
+  private String buildNewStepInstruction(RouteProgress progress) {
+    return buildDefaultFormatInstruction(progress);
+  }
+
+  private String buildImminentInstruction(RouteProgress progress) {
+    return buildContinueFormatInstruction(progress);
+  }
+
+  private String buildUrgentInstruction(RouteProgress progress) {
+    String urgentInstruction = buildThenFormatInstruction(progress);
+
+    if (!TextUtils.isEmpty(urgentInstruction)) {
+      return urgentInstruction;
+    } else {
+      return buildDefaultFormatInstruction(progress);
+    }
+  }
+
+  private String buildArrivalInstruction(RouteProgress progress) {
+    String instruction = progress.getCurrentLegProgress().getUpComingStep().getManeuver().getInstruction();
```

Comment:
`getUpComingStep` needs to be checked if null since there's a chance it could be.